OF-2426: Refactor Group implementation

guusdk opened the pull request:
Reduce complexity by centralizing the responsiblity of dispatching events to event listeners and cache updates.

Avoid using the event listener mechanism that's also being provided to users of this API in the implementation itself.

As a byproduct of the latter, this fixes OF-2426: an event dispatched by the implementation no longer re-populates the cache unexpectedly.

Fishbowler commented on xmppserver/src/main/java/org/jivesoftware/openfire/commands/event/GroupMemberAdded.java:
```diff
@@ -71,37 +69,43 @@ public void execute(SessionData sessionData, Element command) {
-        Map<String, Object> params = null;
-        
-        try {
+        final String wasAdminValue = get(data, "wasAdmin", 0);
+        if (wasAdminValue == null) {
+            note.addAttribute("type", "error");
+            note.setText("WasAdmin required parameter.");
+            return;
+        }
+        final boolean wasAdmin = "1".equals(wasAdminValue) || Boolean.parseBoolean(wasAdminValue);
```
What's populating the "wasAdmin"?

guusdk replied:
A new data field in the form that is passed into this method by the `data` variable.

guusdk commented on xmppserver/src/main/java/org/jivesoftware/openfire/group/GroupManager.java:
```diff
@@ -789,6 +625,423 @@ private void evictCachedUserSharedGroups() {
+        }
+
+        // Update caches.
+        clearGroupNameCache();
```
Possibly, but much of the code follows the "if anything changed, clear all caches." Although that leaves room for improvement, it's a whole lot easier to work with, when diagnosing issues. And this code has issues...

GregDThomas commented on xmppserver/src/main/java/org/jivesoftware/openfire/commands/event/GroupAdminRemoved.java:
```diff
@@ -71,37 +69,35 @@ public void execute(SessionData sessionData, Element command) {
         } catch (GroupNotFoundException e) {
             note.addAttribute("type", "error");
             note.setText("Group not found.");
+            return;
```
"Ditto" - IYSWIM

GregDThomas commented on xmppserver/src/main/java/org/jivesoftware/openfire/commands/event/GroupMemberRemoved.java:
```diff
@@ -71,37 +69,35 @@ public void execute(SessionData sessionData, Element command) {
+            String memberValue = get(data, "member", 0);
+            member = new JID(memberValue);
         }
         catch (NullPointerException npe) {
```
Catching a NPE again

GregDThomas commented on xmppserver/src/main/java/org/jivesoftware/openfire/commands/event/GroupMemberRemoved.java:
```diff
@@ -71,37 +69,35 @@ public void execute(SessionData sessionData, Element command) {
         } catch (GroupNotFoundException e) {
             note.addAttribute("type", "error");
             note.setText("Group not found.");
+            return;
```
duplicated code

GregDThomas commented on xmppserver/src/main/java/org/jivesoftware/openfire/commands/event/GroupModified.java:
```diff
@@ -82,56 +80,92 @@ public void execute(SessionData sessionData, Element command) {
         } catch (GroupNotFoundException e) {
             note.addAttribute("type", "error");
             note.setText("Group not found.");
+            return;
```
duplicated code

GregDThomas commented on xmppserver/src/main/java/org/jivesoftware/openfire/group/DefaultGroupPropertyMap.java:
```diff
@@ -1,19 +1,27 @@
-import java.util.Map;
-import java.util.Map.Entry;
-import java.util.Set;
+import java.util.*;
```
Style thing, I know, but I don't like wildcard imports.

guusdk replied:
I am not bothered by them in the slightest. :) I can count the number of times that these actually caused an issue for me on the fingers of one hand that belongs to the worst employee of a saw mill.

GregDThomas replied:
That's fine. It has caused me problems in the past, which is why I try and avoid them.

(At the time, we had one person using IntelliJ and another with Eclipse. From memory Eclipse imported a wildcard which tripped up IntelliJ because of two different classes with the same name (or vice-versa? I forget)

GregDThomas commented on xmppserver/src/main/java/org/jivesoftware/openfire/group/DefaultGroupPropertyMap.java:
```diff
@@ -32,7 +40,7 @@
 public class DefaultGroupPropertyMap<K,V> extends PersistableMap<K,V> {
 
-    private static final long serialVersionUID = 3128889631577167040L;
+    private static final long serialVersionUID = -7501088917416583815L;
```
Could this change in serialVersionUID cause problems in mixed-version clusters, or during upgrade procedures?

guusdk replied:
I'm not sure, but I don't think that the change is needed at all. The composition of the object did not change. I'll revert this.

GregDThomas commented on xmppserver/src/main/java/org/jivesoftware/openfire/group/DefaultGroupPropertyMap.java:
```diff
@@ -94,17 +102,28 @@ public V put(K key, V value) {
         super.clear();
-        deleteAllProperties();
+
+        // Ignore all entries that are not strings.
```
Might be worth clarifying /why/ you are ignoring them

GregDThomas commented on xmppserver/src/main/java/org/jivesoftware/openfire/group/DefaultGroupPropertyMap.java:
```diff
@@ -191,7 +210,7 @@ public K next() {
             delegate.remove();
             if (current instanceof String) {
-                deleteProperty((String)current);
+                deleteProperty((String)current, null); // FIXME by not providing the original value, some event handlers may not be able to handle this.
```
Does this FIXME need a JIRA reference?

guusdk replied:
Ack. I've raised OF-2430 for this and modified the comment.

GregDThomas commented on xmppserver/src/main/java/org/jivesoftware/openfire/group/DefaultGroupPropertyMap.java:
```diff
@@ -242,8 +261,14 @@ public boolean remove(Object o) {
+            }
             delegate.clear();
-            deleteAllProperties();
+            deleteAllProperties(originalMap);
```
This method doesn't do what the Javadoc says ("Removes all the elements in the set") - possibly "Removes all the elements in the set whose values are Strings" ?

guusdk replied:
Actually, they _are_ all deleted. The map that's passed in is only used for event dispatching. I'm unsure if there is any sense in having this be a generic map - if we're not using Strings, this implementation seems to choke.

GregDThomas commented on xmppserver/src/main/java/org/jivesoftware/openfire/group/DefaultGroupPropertyMap.java:
```diff
@@ -389,15 +414,16 @@ public boolean hasNext() {
+            V value = current.getValue();
             if (key instanceof String) {
-                deleteProperty((String)key);
+                deleteProperty((String)key, value instanceof String ? (String)value : null);
```
(there's clearly a lot of String protection going on here that I don't understand)

guusdk replied:
This implementation apparently only works with Strings, while it accepts a generic type. Something to dig into another time...

GregDThomas commented on xmppserver/src/main/java/org/jivesoftware/openfire/group/GroupManager.java:
```diff
@@ -789,6 +625,423 @@ private void evictCachedUserSharedGroups() {
+            // By forcing the lookup, the cache will automatically receive the required refresh.
+            updatedGroup = getGroup(group.getName(), true);
+        } catch (GroupNotFoundException e) {
+            Log.error("Group '{}' was not found after a member was added to it. This is indicative of a bug in Openfire. Please consider reporting it.", group.getName(), e);
```
This might be also just a co-incidence, if someone deletes the group at about the same time someone is added to it.

Is it possible to single thread some of these actions? Adding a member requires a read lock, deleting the group a read-write lock? Suspect it's tricky across the cluster.

Probably more effort than it's worth.

guusdk replied:
I considered that, but given the complexity of the code, and the way how groups are nested, possibly retrieved from external sources (AD/LDAP), and used all over the place, and clustering, as you mentioned, my fear of introducing deadlocks won the argument.

GregDThomas commented on xmppserver/src/main/java/org/jivesoftware/openfire/group/GroupManager.java:
```diff
@@ -789,6 +625,423 @@ private void evictCachedUserSharedGroups() {
+    private void propertyChangePostProcess(final Group group, final String key, final String originalValue)
+    {
+        switch (key) {
+            case "sharedRoster.showInRoster": {
```
Should these magic Strings be constants so that typo's don't creep in?

* "sharedRoster.showInRoster"
* "everybody"
* "sharedRoster.groupList"

GregDThomas commented on xmppserver/src/test/java/org/jivesoftware/openfire/group/DefaultGroupProviderTest.java:
```diff
@@ -92,11 +91,16 @@ public void tearDown() throws Exception {
         // TODO: this ideally goes in a static @AfterClass method, but that's not supported in JUnit 3.
         for (String fieldName : Arrays.asList("INSTANCE", "provider")) {
-            Field field = GroupManager.class.getDeclaredField(fieldName);
+            final Field field = GroupManager.class.getDeclaredField(fieldName);
```
Yuck. Is there way to avoid this? Perhaps a package-protected method to set the listeners, to be used for test purposes?

guusdk replied:
That's trading in one yuck for another yuck. The test code is a bit of a mess already (with regards to not using `@AfterClass` and the concessions that it needs to do for that). I'm opting to keep the yuckiness in one class.

GregDThomas reviewed (approved):
A good improvement, I think. I've commented on a few picky style issues, but nothing of any real substance I think.

guusdk commented:
Thanks Greg! I believe I've addressed most of your concerns, and choose to not act on a couple of them. Notably, the code duplication in the ad hoc commands. Although there is some room for improvement, I think it'd hurt readability more than that it brings benefit.
